lua/rustaceanvim/neotest/init.lua

```diff
@@ -44,7 +44,23 @@ end
 ---@param file_path string
 ---@return boolean
 NeotestAdapter.is_test_file = function(file_path)
-  return vim.endswith(file_path, '.rs')
+  if file_path:find('target/%a+/build') or not vim.endswith(file_path, '.rs') then
+    return false
+  end
+
+  local content = lib.files.read(file_path)
+  local test_attributes = {
+    ['test'] = true,
+    ['rstest'] = true,
+    ['rstest::rstest'] = true,
+  }
+  for attr in content:gmatch('#%[([%w_:]+)%]') do
+    if test_attributes[attr] then
+      return true
+    end
+  end
```

Comment:
issue: We shouldn't try to determine if the file has a test module, as rust-analyzer is better at doing so than we are.
This limits the amount of test frameworks we can support.

[According to the neotest readme](https://github.com/nvim-neotest/neotest?tab=readme-ov-file#parsing-tests):

> it's OK if a test file has no actual tests in it

Reply:
I agree that rust-analyzer would be much better at determining whether a file contains tests. I'd even tried to use rust-analyzer in the `is_test_file` function, but it appears that the function might be called too early in its initialization, or it was too slow and timing out, as neotest would say that no tests were found.

I took a look at a couple other neotest adapters (neotest-dotnet and neotest-rust) and they both appear to be checking for test attributes in `is_test_file`.

While neotest certainly doesn't mind that you return true in `is_test_file` for files without any actual tests, it does end up polluting the neotest summary sidebar with every single rust file in your project directory, making it hard to find what files actually have tests.

Perhaps a solution could be to check only for the `#[test]` attribute by default, and provide the option for the user to specify additional attributes to be matched for test files.

Reply:
The issue is that a Lua `gmatch` isn't reliable, so it may report false negatives (it has done so in some modules that I tested your fork with). And we can't account for all test frameworks that rust-analyzer might be able to pick up (now or in the future).

I'd like rust-analyzer to be the single source of truth when it comes to detecting test positions.

> it does end up polluting the neotest summary sidebar with every single rust file in your project directory, making it hard to find what files actually have tests.

I think this should be fixed on the neotest side, not on the adapter side.
Either neotest shouldn't include tests that don't have any positions in its summary sidebar, or it should have an API for adapters to tell it not to do so.
I'm sure this is something we can work out with @rcarriga.

Reply:
> The issue is that using a Lua gmatch isn't reliable, so it may report false negatives (it has done so in some modules that I tested your fork with). 

As in it reported matches that it shouldn't have, or reported no matches when it should have? I'd be interested just to know where it failed.

> I'd like rust-analyzer to be the single source of truth when it comes to detecting test positions.

That's reasonable and should result in the most robust solution.

> I think this should be fixed on the neotest side, not on the adapter side.
Either neotest shouldn't include tests that don't have any positions in its summary sidebar, or it should have an API for adapters to tell it not to do so.

I agree that this would be better fixed in neotest rather than in the adapter. I had just followed the example of some other adapters as I wasn't familiar with neotest internals at all.

An adapter/user option to hide files without any test positions would definitely be ideal.

I'll go ahead and revert this function back to simply matching on `.rs` files.

Reply:
> As in it reported matches that it shouldn't have, or reported no matches when it should have? I'd be interested just to know where it failed.

It reported no matches when it should have. I'm not sure why - it was a `tokio::test`.  It reported matches in one file, but not in the other.